### ISSUE TITLE
ADVANCED compilation mode produces code in wrong execution order

### ISSUE THREAD

**Issue by codedhead** [google/closure-compiler#4115]
I found this bug when compiling a third-party library, below is a simplified version to reproduce the bug:

input.js
```
function substr(value, begin, end) { return value.slice(begin, end) }
window.testFunc = function (a, b, c) {
  return substr(a, b+1, b=c);
}
```

command
```
google-closure-compiler -O ADVANCED input.js --js_output_file output.js
```

output.js
```
window.testFunc=function(c,b,a){a=b=a;return c.slice(b+1,a)};
```

The compiler somehow decided to evaluate the third argument `b=c` before the second argument `b+1`, causing the generated code to output incorrect value
```
testFunc('abc',-1,3) // output: '', expected: 'abc'
```

**Comment by rishipal** [comment]
We choose option 2 from the above comment as we don't do good enough side effect tracking to figure out if side-effectable state is actually updated by good side effects, so special casing specific code patterns seems complicated.

**Comment by rishipal** [comment]
Option 2 requires adding / updating unit tests in the following files. I'm currently updating the expected code sections of unit tests to match the results with this change. 

//third_party/java_src/jscomp/javatests/com/google/javascript/jscomp:FunctionArgumentInjectorTest
//third_party/java_src/jscomp/javatests/com/google/javascript/jscomp:FunctionInjectorTest
//third_party/java_src/jscomp/javatests/com/google/javascript/jscomp:InlineFunctionsTest
//third_party/java_src/jscomp/javatests/com/google/javascript/jscomp:ProcessCommonJSModulesTest

Once the tests are updated, I will send out the change for internal code review and update again.

**Comment by rishipal** [comment]
Updated the 4 JSCompiler tests. JSCompiler tests are all passing with the option 2 fix now. 
Started google3-wide presubmit testing for the fix to see what breaks.

**Comment by rishipal** [comment]
Subset of tests (Sample tests) passed. Sent the fix (option 2) for internal code review. google3 presubmit tests still running.

**Comment by rishipal** [comment]
I've submitted the fix internally and it should get auto-pushed to GitHub by copybara in a few ~hours.

**Comment by rishipal** [comment]
This is fixed with bebf834e32d61a14baa608bc49b4081ce0b1ecbc.
